accepted/2020/or-greater-defines/or-greater-defines.md

```diff
@@ -0,0 +1,194 @@
+    #elif NET
+        Net5OrHigherBehavior();
+    #else
+        #error Unhandled TFM
+    #endif
+}
+```
+
+Please note that order is important here. You first handle all the legacy TFMs
+and then you handle the new behavior. To avoid breaking that code once you add
+target .NET 6, you use the versionless TFM for .NET 5 and higher platforms
+(`NET`). However, the code still needs to be revised when you add lower versions
+of existing TFMs.
+
+With this feature, you'll be able to express it like this, which is a lot more
+intuitive:
+
+```C#
+void M()
+{
+    #if NET5_0_OR_GREATER
+        Net5OrHigherBehavior();
+    #elif NETCOREAPP3_1_OR_GREATER
+        LegacyNetCoreBehavior();
+    #elif NET461_OR_GREATER
+        LegacyNetFxBehavior();
+    #else
+        #error Unhandled TFM
+    #endif
+}
+```
+
+The idea is to start from newest to oldest.
+
+This also makes the code more resilient for cases where you add new frameworks
+with lower versions later. For example, if you start targeting .NET Core 3.0,
+the former code would go down the path for .NET 6 while the latter will fail
+with "Unhandled TFM", which is a lot more logical for the person having to
+maintain the code.
+
+## Scenarios and User Experience
+
+### Multi-targeting for devices to provide a cross-platform API
+
+Joletta is building a cross-platform library for geolocation. Since there is no
+built-in API for accessing geolocation, she needs to call different APIs for
+different platforms, so she uses `#if` to write different code, depending on the
+OS she is building for. Since she needs to provide custom implementations per
+OS, she wants to fail whenever a new TFM is added.
+
+```C#
+public static class Gps
+{
+    public static GpsCoordinate GetCoordinates()
+    {
+        #if NETSTANDARD2_0
+            throw new PlatformNotSupportedException();
+        #elif WINDOWS7_0
+            return CallWindowsApi();
+        #elif ANDROID14_0
+            return CallAndroidApi();
+        #elif IOS11_0
+            return CallIOSApi();
+        #else
+            #error Provide implementation
+        #endif
+    }
+}
+```
+
+### When running on a later version of .NET, use a newer API
+
+Zachariah is building a .NET imaging library. On .NET 5, we have added enough
+hardware intrinsics so that he can implement the blur functionality in a much
+faster way. However, he still wants to support .NET Standard 2.0, so he decides
+to use `#if` to multi-target for .NET 5.
+
+Since he checks for the minimum version of the framework that introduced the
+API, adding a new version wouldn't change this code, so he simply puts the
+fallback logic for the slow version in the `#else` branch, with no call to
+`#error`.
+
+```C#
+public class Picture
+{
+    public void Blur(float radius)
+    {
+        #if NET5_0_OR_HIGHER
+            FastBlurUsingHardwareIntrinsics();
+        #else
+            SlowBlurUsingSimpleMath();
+        #endif
+    }
+}
+```
+
+## Requirements
```

Comment:
Do the requirements include out-of-support TFMs❔ If not, suggest adding that as a non-goal.

Reply:
The requirements section includes the list of frameworks for which we will do the defines, so I think that's called out.

Reply:
See my response [here](https://github.com/dotnet/designs/pull/164#issuecomment-723326659). Unless there is a significant cost increase, I think it's easier if do this for all TFMs that can be targeted from the .NET 5 SDK.